velox/exec/Window.cpp

```diff
@@ -0,0 +1,465 @@
+    peerStartBuffer->setSize(numRows);
+    peerEndBuffer->setSize(numRows);
+    frameStartBuffer->setSize(numRows);
+    frameEndBuffer->setSize(numRows);
+    allFuncsPeerStartBuffer[w] = peerStartBuffer;
+    allFuncsPeerEndBuffer[w] = peerEndBuffer;
+    allFuncsFrameStartBuffer[w] = frameStartBuffer;
+    allFuncsFrameEndBuffer[w] = frameEndBuffer;
+
+    auto rawPeerStartBuffer = peerStartBuffer->asMutable<size_t>();
+    auto rawPeerEndBuffer = peerEndBuffer->asMutable<size_t>();
+    auto rawFrameStartBuffer = frameStartBuffer->asMutable<size_t>();
+    auto rawFrameEndBuffer = frameEndBuffer->asMutable<size_t>();
+    allFuncsRawPeerStartBuffer[w] = rawPeerStartBuffer;
+    allFuncsRawPeerEndBuffer[w] = rawPeerEndBuffer;
+    allFuncsRawFrameStartBuffer[w] = rawFrameStartBuffer;
+    allFuncsRawFrameEndBuffer[w] = rawFrameEndBuffer;
+  }
+
+  size_t firstPartitionRow = partitionStartRows_[currentPartitionIndex_];
+  size_t lastPartitionRow = partitionStartRows_[currentPartitionIndex_ + 1] - 1;
+  for (int i = startRow, j = 0; i < endRow; i++, j++) {
+    // Compute the next peerStart and peerEnd rows (if this is the first row
+    // of the partition or if we are past previous peerGroup).
+    if (i == firstPartitionRow || i >= peerEndRow_) {
+      peerStartRow_ = i;
+      peerEndRow_ = i;
+      while (peerEndRow_ <= lastPartitionRow) {
+        if (peerCompare(
+                returningRows_[peerStartRow_], returningRows_[peerEndRow_])) {
+          break;
+        }
+        peerEndRow_++;
+      }
+    }
+    for (int w = 0; w < numFuncs; w++) {
+      // The peer and frame values set in the WindowFunction::apply buffers
+      // are the offsets within the current partition, so offset all row
+      // indexes correctly.
+      allFuncsRawPeerStartBuffer[w][j] = peerStartRow_ - firstPartitionRow;
+      allFuncsRawPeerEndBuffer[w][j] = peerEndRow_ - 1 - firstPartitionRow;
+
+      auto frameEndPoints = findFrameEndPoints(
+          w, partitionStartRows_[currentPartitionIndex_], endRow, i);
+      VELOX_CHECK_EQ(frameEndPoints.second, i);
+      allFuncsRawFrameStartBuffer[w][j] =
+          frameEndPoints.first - firstPartitionRow;
+      allFuncsRawFrameEndBuffer[w][j] =
+          frameEndPoints.second - firstPartitionRow;
+    }
+  }
+  // Invoke the apply method for the WindowFunctions
+  for (int w = 0; w < numFuncs; w++) {
+    windowFunctions_[w]->apply(
+        allFuncsPeerStartBuffer[w],
+        allFuncsPeerEndBuffer[w],
+        allFuncsFrameStartBuffer[w],
+        allFuncsFrameEndBuffer[w],
+        startRow - firstPartitionRow,
+        bufferIndex,
+        windowFunctionOutputs[w]);
+  }
+
+  numRowsApplied_ += numRows;
+  if (endRow == partitionStartRows_[currentPartitionIndex_ + 1]) {
+    currentPartitionIndex_++;
+  }
+}
+
+RowVectorPtr Window::getOutput() {
+  if (finished_ || !noMoreInput_ || returningRows_.size() == numRowsApplied_) {
```

Comment:
Can we remove 'returningRows_.size() == numRowsApplied_' check by adding logic to set finished to true when we are done processing?

Reply:
That existed already. Removing this condition.